lightning-c-bindings/README.md

```diff
@@ -0,0 +1,193 @@
+   Note that calling `X_free` will do nothing if the underlying pointer is NULL or if the reference
+   flag is set.
+
+   You MUST NOT create such wrapper structs manually, relying instead on constructors which have
+   been mapped from equivalent Rust constructors.
+
+   Note that, thanks to the is-reference flag and the pointer being NULLable, such structs
+   effectively represent both `&RustThing`, `RustThing`, and `Option<RustThing>`. Check the
+   corresponding Rust documentation for the function or struct you are using to ensure you use
+   the correct call semantics.
+
+   For example, this is the mapping of ChannelManager.
+   ```
+   typedef struct MUST_USE_STRUCT LDKChannelManager {
+      /** ... */
+      LDKlnChannelManager *inner;
+      bool _underlying_ref;
+   } LDKChannelManager;
+   ```
+
+ * Traits are mapped into a concrete struct containing a void pointer (named `this_arg` and a jump
+   table listing the functions which the trait must implement. The void pointer may be set to any
+   value and is never interpreted (or dereferenced) by the bindings logic in any way. It is passed
+   as the first argument to all function calls in the trait. You may wish to use it as a pointer to
+   your own internal data structure, though it may also occasionally make sense to e.g. cast a file
+   descriptor into a void pointer and use it to track a socket.
+
+   This should remind you of a C++ vtable, only written out by hand and with the class only
+   containing a pointer, instead of the regular class data.
+
+   Each trait additionally contains a `free` and `clone` function pointer, which may be NULL. The
+   `free` function is passed the void pointer when the object is `Drop`ed in Rust. The `clone`
+   function is passed the void pointer when the object is `Clone`ed in Rust, returning a new void
+   pointer for the new object. If the `free` pointer is NULL, you will not receive any notification
+   when the trait is no longer needed. If the `clone` pointer is NULL, we assume that the trait
+   object may be `memcpy()`'d to create copies. Note that if you release resources with `free`
+   without implementing `clone`, you will likely end up with use-after-free bugs.
+
+   For example, `LDKSocketDescriptor` is mapped as follows:
+   ```
+   typedef struct LDKSocketDescriptor {
+      void *this_arg;
+      /** ... */
+      uintptr_t (*send_data)(void *this_arg, LDKu8slice data, bool resume_read);
+      /** ... */
+      void (*disconnect_socket)(void *this_arg);
+      bool (*eq)(const void *this_arg, const void *other_arg);
+      uint64_t (*hash)(const void *this_arg);
+      void *(*clone)(const void *this_arg);
+      void (*free)(void *this_arg);
+   } LDKSocketDescriptor;
+   ```
+
+ * Rust structs that implement a trait result in the generation of an `X_as_Y` function which allows
+   you to use the native Rust object in place of the trait. Such generated objects are only valid as
```

Comment:
Hmm I found this a bit confusing.. Maybe instead: "Rust structs that implement a trait result in the generation of an `X_as_Y` function, which takes in a C struct wrapping the Rust native object and returns a generated trait object. Such generated objects are only valid as long as the original Rust native object has not been `free`'d or moved as part of a Rust function call. For example: -insert e.g. KeysManager_as_KeysInterface example-